### PR TITLE
Distributed N-dimensional sort

### PR DESCRIPTION
**Pull request by mfoerste4**
This PR adds true multi GPU support for sort (and partition) of multidimensional arrays.

There are still several scalability issues which are being addressed one by one, but the code feels stable and fast enough to get merged.

* I removed the 1d distributed code and merged it into the more general n-d one
* At this point all data is re-balanced after sorting, meaning that no unbound stores are used. This keeps it simple for now but it might change later to optimize for 1D
* UPDATE: 1D data is *not* re-balanced anymore

### CODE REVIEW

File: src/cunumeric/sort/sort.cu [hunk 1]
```diff
@@ -95,338 +137,1460 @@ struct SortPiece {
...
+        copy_data_to_merge_buffers<<<grid_shape, block_shape, 0, stream>>>(segment_blocks,
+                                                                           size_send,
+                                                                           local_sorted.indices,
+                                                                           idc_send_buffers_ptr,
+                                                                           num_segments_l,
+                                                                           num_segments_l_aligned,
+                                                                           segment_size_l,
+                                                                           my_rank,
+                                                                           num_sort_ranks);
+        CHECK_CUDA(cudaStreamSynchronize(stream));  // needed before Z-copy destroy()?
```
Comment (magnatelee) [inline src/cunumeric/sort/sort.cu:1454]: I'd make sure if this synchronization is really necessary. (same applies to other synchronization calls too)

**Comment by magnatelee** [comment]
Made a first pass. I skimmed through the code except the rebalancing part, which I'll take a look tomorrow. the code is fairly technical and at least I couldn't find anything obviously wrong. @mfoerste4 are you confident that the current test is sufficient testing new N-D implementation? I guess adding a couple more tests with small inputs would be helpful to check if the corner cases are handled correctly.

File: cunumeric/sort.py [hunk 2]
```diff
@@ -59,34 +59,25 @@ def sort_swapped(output, input, argsort, sort_axis, stable):
 def sort_task(output, input, argsort, stable):
     task = output.context.create_task(CuNumericOpCode.SORT)
 
-    needs_unbound_output = output.runtime.num_gpus > 1 and input.ndim == 1
-
-    if needs_unbound_output:
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply cunumeric/sort.py:64]: Yes, unfortunately the overhead of re-balancing is too high to neglect this.

File: src/cunumeric/sort/sort.cu [hunk 3]
```diff
@@ -95,338 +137,1460 @@ struct SortPiece {
...
+    auto worker_count_d = create_buffer<int32_t>(1, Memory::GPU_FB_MEM);
+    int worker_count[1] = {(segment_size_l > 0 ? 1 : 0)};
+    CHECK_CUDA(cudaMemcpyAsync(
+      worker_count_d.ptr(0), &worker_count[0], sizeof(int32_t), cudaMemcpyHostToDevice, stream));
+    CHECK_NCCL(ncclAllReduce(
+      worker_count_d.ptr(0), worker_count_d.ptr(0), 1, ncclInt32, ncclSum, *comm, stream));
+    CHECK_CUDA(cudaMemcpyAsync(
+      &worker_count[0], worker_count_d.ptr(0), sizeof(int32_t), cudaMemcpyDeviceToHost, stream));
+    CHECK_CUDA(cudaStreamSynchronize(stream));
+    if (worker_count[0] < num_ranks) {
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu]: Yes, there are still some unwritten dependencies here. In case we want to allow more dynamic resource assignment we would need to provide some more information on it to the tasks I suppose.

File: src/cunumeric/sort/sort.cu [hunk 4]
```diff
@@ -95,338 +137,1460 @@ struct SortPiece {
...
+                               stream));
+
+    auto recv_buffer =
+      create_buffer<SegmentSample<VAL>>(aligned_count * num_sort_ranks, Memory::GPU_FB_MEM);
+
+    CHECK_NCCL(ncclGroupStart());
+    for (size_t r = 0; r < num_sort_ranks; r++) {
+      if (r != my_sort_rank) {
+        CHECK_NCCL(ncclSend(send_buffer.ptr(0),
+                            num_samples_l * sizeof(SegmentSample<VAL>),
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu]: I adjusted the transfer size for several communication blocks. The next tests will show whether it is still needed. It wont hurt for sure.

File: src/cunumeric/sort/sort.cu (hunk 1 above)
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu]: I removed some of the synchronization calls that were outdated. The 2 synchronizations related to Z-copy (especially the one in question here) are mandatory - the tests fail right away when I remove them.

**Comment by mfoerste4** [comment]
Thanks for your suggestions @magnatelee . Regarding the test coverage I usually run the test suite with 1-n GPUs which covers a whole bunch of edge cases where the partition is done across multiple axes. What corner cases are you referring to? I would be happy to add more tests.

File: src/cunumeric/sort/sort.cu [hunk 5]
```diff
@@ -87,6 +92,43 @@ auto get_16b_aligned_count = [](auto count, auto element_bytes) {
...
+  size_t result = 1;
+  while (result < THREADS_PER_BLOCK &&
+         (segment_size_l * result < THREADS_PER_BLOCK ||
+          (segment_size_l * result <= THREADS_PER_BLOCK * 16 && num_segments_l / result > 256))) {
+    result *= 2;
+  }
+  return result;
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply src/cunumeric/sort/sort.cu:113]: I was talking about the lines separating sections. We don't have to change them, but it's weird that you had to put 10 lines.

File: src/cunumeric/sort/sort.cu [hunk 6]
```diff
@@ -95,277 +137,1514 @@ struct SortPiece {
...
-                                 stream));
-      pos += size_send[r];
+    {
+      Buffer<VAL*> val_send_buffers_ptr = create_buffer<VAL*>(num_sort_ranks, Memory::Z_COPY_MEM);
+      for (size_t r = 0; r < num_sort_ranks; r++) {
+        val_send_buffers_ptr[r] = val_send_buffers[r].ptr(0);
+      }
+
+      // is either 1, 2, 4, ... , THREADS_PER_BLOCK/2
+      int segments_per_threadblock =
```
Comment (magnatelee) [inline src/cunumeric/sort/sort.cu:1527]: Lines 1527-1533 seem to repeat in several places. I think you extract them out to a function.

File: src/cunumeric/sort/sort.cu [hunk 7]
```diff
@@ -95,277 +137,1514 @@ struct SortPiece {
...
+        compute_segments_per_threadblock(segment_size_l, num_segments_l);
+      dim3 block_shape =
+        dim3(THREADS_PER_BLOCK / segments_per_threadblock, segments_per_threadblock);
+      const size_t num_blocks_y = (num_segments_l + block_shape.y - 1) / block_shape.y;
+      const size_t num_blocks_x = ((segment_size_l + 31) / 32 + block_shape.x - 1) / block_shape.x;
```
Comment (magnatelee) [inline src/cunumeric/sort/sort.cu:1532]: can you leave a comment on the purpose of this heuristic?

**Review by magnatelee — APPROVED** [review]

**Comment by magnatelee** [comment]
> Thanks for your suggestions @magnatelee . Regarding the test coverage I usually run the test suite with 1-n GPUs which covers a whole bunch of edge cases where the partition is done across multiple axes. What corner cases are you referring to? I would be happy to add more tests.

bad things usually happen when small inputs are partitioned across multiple GPUs in a way that some GPUs get empty pieces. If you already check such cases with the current integration test, I guess we don't need any extra tests.

File: src/cunumeric/sort/sort.cu (hunk 7 above)
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu:1532]: The 32 does not refer to the warp size. It is the maximum number of elements every thread is responsible for. In case there is more than (32*threads) elements to be handled in a segment the amount of blocks assigned per segment will be increased.

This should only trigger in case we have large segments (extreme case is 1D). Here I want to have multiple blocks working on the data and not just one to saturate memory bandwidth.

Comment (magnatelee) [reply src/cunumeric/sort/sort.cu:1532]: in case where the segment is small, is this still better than doing `(segment_size_l + block_shape.x - 1) / block_shape.x` (i.e., each thread getting assigned to only one element)? I think a magic number like 32 needs an explanation for anyone, including yourself, who would need to touch this code later.

File: src/cunumeric/sort/sort.cu (hunk 6 above)
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu:1527]: I extracted the launchconfig generation into a more generic utility function and added some comments.

File: src/cunumeric/sort/sort.cu [hunk 8]
```diff
@@ -508,30 +1675,48 @@ struct SortImplBody<VariantKind::GPU, CODE, DIM> {
...
       }
+
+      sample_sort_nccl_nd(local_sorted,
+                          output_ptr,
+                          local_rank,
+                          num_ranks,
+                          segment_size_g,
+                          local_rank % num_sort_ranks,
+                          num_sort_ranks,
+                          &sort_ranks[0],
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu]: No, there is no reason for that. I think I had the API first and created the argument afterwards. I adjusted the code to pass `sort_ranks.data()`

File: src/cunumeric/sort/sort.cu (hunk 7 above)
Comment (mfoerste4) [reply src/cunumeric/sort/sort.cu]: In case the segments are small (and we have many of them) the threadblocks should already cover multiple segments which leaves only a fraction of threads per segment. The magic-32 only comes into play once we have large rows and decide that we need multiple blocks per row.
